orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py

```diff
@@ -250,25 +254,62 @@ def _get_exported_model(self, *inputs, **kwargs):
-                torch.onnx.export(self._flattened_module,
-                                  sample_inputs_as_tuple,
-                                  f,
-                                  input_names=self._input_info.names,
-                                  output_names=output_names,
-                                  opset_version=ONNX_OPSET_VERSION,
-                                  do_constant_folding=False,
-                                  training=self._export_mode,
-                                  dynamic_axes=self._input_info.dynamic_axes,
-                                  verbose=self._loglevel < _logger.LogLevel.WARNING,
-                                  export_params=False,
-                                  keep_initializers_as_inputs=True)
+            if self._onnx_export_type == torch.onnx.OperatorExportTypes.ONNX_FALLTHROUGH:
+                cloned_flattened_module = copy.deepcopy(self._flattened_module)
+                with _logger.suppress_os_stream_output(log_level=self._loglevel):
+                    torch.onnx.export(cloned_flattened_module,
+                                    sample_inputs_as_tuple,
+                                    f,
+                                    input_names=self._input_info.names,
+                                    output_names=output_names,
+                                    opset_version=ONNX_OPSET_VERSION,
+                                    do_constant_folding=False,
+                                    training=self._export_mode,
+                                    dynamic_axes=self._input_info.dynamic_axes,
+                                    operator_export_type=self._onnx_export_type,
+                                    custom_opsets={"prim": 1},
+                                    verbose=self._loglevel < _logger.LogLevel.WARNING)
+
+                exported_model = onnx.load_model_from_string(f.getvalue())
+                if len(exported_model.opset_import) > 1:
+                    exported_model.opset_import[1].domain = 'com.microsoft'
+                index = 0
+                for node in exported_model.graph.node:
+                    if node.domain == 'com.microsoft' and node.op_type in ["PythonOp"]:
+                        output_names = list(node.output)
+                        del node.output[:]
+                        node.output.append(output_names[0] + '_ctx')
+                        node.output.extend(output_names)
+                    if not node.name:
+                        # give a name for debugging
+                        node.name = node.op_type + "_id_" + str(index)
+                        index += 1
+
+                for kclass in torch.autograd.Function.__subclasses__():
+                    # Sometimes, we find the same functions multiple times, so we allow repeated
+                    # registrations.
+                    register_torch_autograd_function(kclass.__qualname__, kclass, True)
```

Comment:
Any reason we cant do that on the orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py::_export()?

If this cant be added there, we could create a helper at orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_exporter.py and move this graph modification code to that file. GraphExecutionManager should not worry about editing onnx files

Reply:
[peng]

Reply:
Yeah will move the post process into _custom_autograd_function_exporter.p